synfig-studio/src/gui/dialogs/canvasproperties.h

```diff
@@ -43,39 +53,131 @@
 /* === C L A S S E S & S T R U C T S ======================================= */
 
 namespace Gtk { class TreeView; };
-namespace synfigapp { class CanvasInterface; };
 
 namespace studio
 {
 
 class CanvasProperties  :  public Gtk::Dialog
 {
+	Glib::RefPtr<Gtk::Builder> canvas_properties;
+
 	etl::handle<synfigapp::CanvasInterface> canvas_interface_;
-	Widget_RendDesc widget_rend_desc;
-	Gtk::Entry entry_id;
-	Gtk::Entry entry_name;
-	Gtk::Entry entry_description;
+	synfig::RendDesc rend_desc_;
+	sigc::signal<void> signal_changed_;
+
+	Gtk::Box *outer_box;
+	Gtk::Box *inner_box;
+
+	// Canvas Info
+	Gtk::Entry *entry_id;
+	Gtk::Entry *entry_name;
+	Gtk::Entry *entry_desc;
+
+	// Image
+	Gtk::Label *ratio_label;
+
+	Gtk::SpinButton *entry_width;
+	Gtk::SpinButton *entry_height;
+	Gtk::SpinButton *entry_xres;
+	Gtk::SpinButton *entry_yres;
+	Gtk::SpinButton *entry_phy_w;
+	Gtk::SpinButton *entry_phy_h;
+	Gtk::SpinButton *entry_span;
+
+	studio::Widget_Link *toggle_wh_r;
+	studio::Widget_Link *toggle_res_r;
+
+	studio::Widget_Vector *entry_tl;
+	studio::Widget_Vector *entry_br;
+
+	// Time
+	Gtk::SpinButton *entry_fps;
+
+	studio::Widget_Time *entry_start;
+	studio::Widget_Time *entry_end;
+	studio::Widget_Time *entry_dur;
+
+	// Gamma Correction
+	Gtk::SpinButton *entry_gamma_r;
+	Gtk::SpinButton *entry_gamma_g;
+	Gtk::SpinButton *entry_gamma_b;
+
+	Gtk::Scale *scale_gamma_r;
+	Gtk::Scale *scale_gamma_g;
+	Gtk::Scale *scale_gamma_b;
+
+	studio::Widget_GammaPattern *gamma_pattern;
+
+	// Other
+	Gtk::CheckButton *toggle_pxl_w;
+	Gtk::CheckButton *toggle_pxl_h;
+	Gtk::CheckButton *toggle_pxl_a;
+	Gtk::CheckButton *toggle_img_w;
+	Gtk::CheckButton *toggle_img_h;
+	Gtk::CheckButton *toggle_img_a;
+	Gtk::CheckButton *toggle_img_s;
+
+	studio::Widget_Vector *entry_focus;
 
 	bool dirty_rend_desc;
 
+	mutable int update_lock;
+
+	struct UpdateLock
+	{
+		int &locked;
+		UpdateLock(int &locked):locked(locked){locked++;}
+		~UpdateLock(){locked--;}
+	};
+
```

Comment:
Are these lines useful?

Reply:
I think so... The last time I removed that portion of code synfig crashed when I modified any value in the canvas properties dialog under Notebook tab. I don't remember exactly how.

I'll go back and see what the behavior is again.

Reply:
> synfig crashed when I modified any value in the canvas properties dialog under Notebook tab.

I was wrong. Synfig crashed when launching.

![Screenshot from 2020-10-17 06-22-18](https://user-images.githubusercontent.com/24214509/96324791-2aaf9200-1041-11eb-93eb-6d30bed1d099.png)

Maybe the lock is to prevent the dialog contents to be set more than once based on what I observed while commenting out those lines in widgets signal call
```c++
void
CanvasProperties::on_widget_changed()
{
    if (update_lock) return;
    UpdateLock lock(update_lock);
    ...
}
```